src/odemis/util/spot.py

```diff
@@ -126,86 +128,260 @@ def SpotIntensity(data, background=None):
+        dIdi = scipy.signal.convolve2d(dIdi, k, boundary='symm', mode='same')
+        dIdj = scipy.signal.convolve2d(dIdj, k, boundary='symm', mode='same')
+    dI2 = np.square(dIdi) + np.square(dIdj)
+
+    # Discard entries where the intensity gradient magnitude is zero, flatten
+    # the array in the same go.
+    idx = np.flatnonzero(dI2)
+    ik = np.take(ik, idx)
+    jk = np.take(jk, idx)
+    dIdi = np.take(dIdi, idx)
+    dIdj = np.take(dIdj, idx)
+    dI2 = np.take(dI2, idx)
+
+    # Construct the set of equations for a line passing through the midpoint
+    # (ik, jk), parallel to the gradient intensity, in implicit form:
+    # `a*i + b*j + c = 0`, normalized such that `a^2 + b^2 = 1`.
+    dI = np.sqrt(dI2)
+    a = -dIdj / dI
+    b = dIdi / dI
+    c = a * ik + b * jk
+
+    # Weighting: weight by the square of the gradient magnitude and inverse
+    # distance to the centroid of the square of the gradient intensity
+    # magnitude.
+    sdI2 = np.sum(dI2)
+    i0 = np.sum(dI2 * ik) / sdI2
+    j0 = np.sum(dI2 * jk) / sdI2
+    w = np.sqrt(dI2 / np.hypot(ik - i0, jk - j0))
+
+    # Solve the linear set of equations in a least-squares sense.
+    ic, jc = np.linalg.lstsq(np.vstack((w * a, w * b)).T, w * c)[0]
+
+    # Convert from index (top-left) to (center) position information.
+    xc = jc - 0.5 * float(m) + 0.5
+    yc = ic - 0.5 * float(n) + 0.5
+
+    return xc, yc
+
+
+def sedisk(r=3):
+    """
+    Create a flat disk-shaped structuring element with the specified radius r.
+
+    Parameters
+    ----------
+    r : Non-negative integer. Disk radius (default: 3)
+
+    Returns
+    -------
+    nhood : Structuring element
+    """
+    y, x = np.mgrid[-r:r + 1, -r:r + 1]
+    nhood = (x * x + y * y) <= (r * (r + 1))
+    return nhood
+
+
+def MaximaFind(image, qty, nsize=18):
+    """
+
+    Parameters
+    ----------
+    image : array like
+        Data array containing the greyscale image.
+    qty : int
+        The amount of maxima you want to find.
+    nsize : int
+
+    Returns
+    -------
+    refined_position : array like
+        A 2D array of shape (N, 2) containing the coordinates of the maxima.
+
+    """
+    filtered = BandPassFilter(image, 1, nsize)
+
+    # dilation
+    s = sedisk(nsize // 2)
+    dilated = ndimage.grey_dilation(filtered, footprint=s)
+
+    # find local maxima
+    binary = (filtered == dilated)
+
+    # thresholding
+    qradius = max(nsize // 4, 1)
+    BWdil = ndimage.binary_dilation(binary, structure=sedisk(qradius))
+
+    labels, num_features = ndimage.label(BWdil)
+    index = np.arange(1, num_features + 1)
+    mean = ndimage.mean(filtered, labels=labels, index=index)
+    si = np.argsort(mean)[::-1][:qty]
+    # estimate spot center position
+    pos = np.array(ndimage.center_of_mass(filtered, labels=labels,
+                                          index=index[si]))[:, ::-1]
+    if np.any(np.isnan(pos)):
+        logging.warning('not all positions found')
```

Comment:
That's not a very helpful warning. If an Odemis (GUI) user ever triggers it, he has probably no way to tell what this is about, and no way to influence it.
So I'd lower it to just a debug log, and write it something like: "Only %d maxima found, while expected %d".